Libplanet.Net/Consensus/MessageCache.cs

```diff
@@ -0,0 +1,158 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using Libplanet.Net.Messages;
+
+namespace Libplanet.Net.Consensus
+{
+    /// <summary>
+    /// Cache object that stores recently seen <see cref="Message"/>s.
+    /// </summary>
+    public class MessageCache
+    {
+        private readonly object _lock;
+        private readonly int _len;
+        private readonly int _gossip;
+        private readonly Dictionary<MessageId, Message> _messages;
+        private readonly ImmutableArray<MessageId>[] _history;
```

Comment:
I'm afraid I misunderstood the meaning of your comment. Do you mean `_history` should me `ImmutableArray<MessageId?>`?

Reply:
Actually, I took a mistake reading this code. I thought about cursor based overwriting cache without `Shift()` but since elements of `_history` is immutable, overwriting seemed to be not efficient for memory (for this, not sure).
So I thought about making it mutable, but now I found this will let original instance unsafe, open to mutation...

Since this question was from my mistake, I'll resolve it right now.